### PR TITLE
PLANNER-1838 Conditional propagation documentation

### PR DESCRIPTION
**Pull request by triceo**

### CODE REVIEW

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc [hunk 1]
```diff
@@ -482,6 +483,58 @@ The following example finds a computer which runs the most power-demanding proce
...
+    private Constraint unusedComputer(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(CloudComputer.class)
+                .ifNotExists(CloudProcess.class, Joiners.equal(Function.identity(), CloudProcess::getComputer))
+                .penalize("unusedComputer",
+                        HardSoftScore.ONE_HARD,
+                        computer -> ...);
+    }
+----
+
+Here, only the computers without processes running on them will be penalized.
```
Comment (ge0ffrey) [inline optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc]: future tense: avoid "will" -> "are penalized"

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc [hunk 2]
```diff
@@ -482,6 +483,58 @@ The following example finds a computer which runs the most power-demanding proce
...
+
+Also note the use of the `Joiner` class to limit the constraint matches.
+For a description of `Joiners`, see <<constraintStreamsJoin,joining>>.
+Conditional propagation operates much like joining, with the exception of not increasing the
+<<constraintStreamsCardinality, stream cardinality>>.
+The matches from `ifExists()` and other similar building blocks will not be available further down the stream.
```
Comment (ge0ffrey) [inline optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc]: will not be -> are not

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc [hunk 3]
```diff
@@ -482,6 +483,58 @@ The following example finds a computer which runs the most power-demanding proce
...
+
+[NOTE]
+====
+For performance reasons, using conditional propagation with the appropriate `Joiner` is preferable to joining.
+Joining should only be used in cases where the other instance is actually required for another operation further down
+the stream.
```
Comment (ge0ffrey) [inline optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc]: I'd argue that's not the real difference between join() and ifExists().
This is the real difference:
[a1, a2, a3].join([b1, b2], someCondition) gives up to 6 matches.
[a1, a2, a3].ifExists([b1, b2], someCondition) gives up to 3 matches.

Comment (triceo) [reply optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc]: Well, that's the same thing I'm trying to say. Obviously, I'm not saying it well enough.

**Pull request merged**